### ISSUE TITLE
Tracking issue for #[doc(alias = "...")]

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#50146]
Edit: Implemented in #49757. From the docs:

This feature allows you to add alias(es) to an item when using the `rustdoc` search through the `doc(alias)` attribute. Example:

```rust,no_run
#![feature(doc_alias)]
#[doc(alias = "x")]
#[doc(alias = "big")]
pub struct BigX;
```

Then, when looking for it through the `rustdoc` search, if you enter "x" or "big", search will show the `BigX` struct first.

**Comment by frewsxcv** [comment]
which pull request was this implemented in?

**Comment by GuillaumeGomez** [comment]
[woups...](https://github.com/rust-lang/rust/pull/49757)

**Comment by GuillaumeGomez** [comment]
Since it's been implemented a while ago, I'm closing it.

**Comment by petrochenkov** [comment]
The feature is still unstable and this is the tracking issue, so the compiler points here.
Reopening.

**Comment by sgrif** [comment]
I tried using this feature today, it doesn't appear to actually be implemented. I tried adding `#[doc(alias = "...")]` to multiple items, none of them showed up when searching for the alias, nor did they get added to `search-index.js` (where I assume they were supposed to live)

**Comment by sgrif** [comment]
Hm, it does appear to work for structs at least -- but not all items that otherwise appear in search. In particular it doesn't appear on trait functions

**Comment by GuillaumeGomez** [comment]
Ah? Interesting. I'll need to take a look at what's generated.

**Comment by GuillaumeGomez** [comment]
I propose to stabilize this feature in order to make it usable outside of the std docs because of a few cases:
 * crates with proc-macro can have their own syntax and might want to document it
 * "overload" some items naming (for example in gtk-rs, since one of its dependencies has a `Window` type, we had to rename it `GtkWindow` but it'd make sense to have it with better ranking when looking for `Window` in the docs).

As for now, I don't see anything that could be an obstacle to stabilizing this feature?

@rfcbot fcp merge

**Comment by GuillaumeGomez** [comment]
@rfcbot fcp merge

**Comment by ollie27** [comment]
Sorry about not getting back to this sooner.

What are some concrete use-cases for this? The `GtkWindowExt` example doesn't seem very compelling because searching for [`WindowExt`](https://gtk-rs.org/docs/gtk/?search=WindowExt) already brings up `GtkWindowExt`.

@rfcbot concern use-cases

The name "alias" alone doesn't make it clear this affects only search. I can't think of anything better than `#[doc(search_alias = "...")]` or maybe `#[doc(search_keyword = "...")]` though.

@rfcbot concern naming

As pointed out by https://github.com/rust-lang/rust/issues/50146#issuecomment-496602762 this doesn't yet work for associated items.

@rfcbot concern associated-items

Currently rustdoc is silently accepting things like:

```rust
#[doc(alias)]
#[doc(alias(0))]
#[doc(alias("foo"))]
#[doc(alias = 0)]
```

They have no effect but should of course be hard errors.

@rfcbot concern invalid-forms

**Comment by GuillaumeGomez** [comment]
> The name "alias" alone doesn't make it clear this affects only search. I can't think of anything better than `#[doc(search_alias = "...")]` or maybe `#[doc(search_keyword = "...")]` though.

This is a fair point. I personally like the current one but if another wording seems better, all good for me (even though I don't like much "search_keyword").

> Currently rustdoc is silently accepting things like:

I intended to add a lint for the case where the alias was discarded, which seemed like an idea solution. So I guess it'd wrap this as well.

> As pointed out by #50146 (comment) this doesn't yet work for associated items.

#71724 should fix this problem.

> What are some concrete use-cases for this? The `GtkWindowExt` example doesn't seem very compelling because searching for `WindowExt` already brings up `GtkWindowExt`.

No, it wasn't the best example. Here a better one: I discovered recently that adding aliases for C library bindings would be very helpful (which was my original point, but here's why). For example, you type the name of the C function and the alias shows you the rust equivalent. It'd be a big plus instead of having to grep through the source code.

**Comment by ollie27** [comment]
> For example, you type the name of the C function and the alias shows you the rust equivalent. It'd be a big plus instead of having to grep through the source code.

Could you provide some links to crates which would benefit from this along the search terms that aren't producing useful results yet?

**Comment by lzutao** [comment]
It seems that this feature doesn't work with trait methods.

**Comment by GuillaumeGomez** [comment]
@ollie27 Litterally all C bindings. Sometimes, to improve the API, you have to rename things and call a few functions from the C library inside a same function. Or just the renaming might be confusing sometimes (it happened a lot for me when using `rust-sdl2`).

**Comment by GuillaumeGomez** [comment]
@lzutao doc alias [tests](https://github.com/rust-lang/rust/blob/c4d9318be6a493da4c6aa307dd4de7e24a15120a/src/test/rustdoc-js/doc-alias.rs#L22-L23) seem to say the opposite. ;)

Maybe just try to update your nightly version?

**Comment by lzutao** [comment]
I've checked this feature again. It kinda works but I had to enter the whole (exact) aliasing keywords
for it to display in search board.

* If entering an extract alias:
![image](https://user-images.githubusercontent.com/15225902/83355260-83c7a380-a388-11ea-99fb-3e2bb59a9e77.png)

* If not:
![image](https://user-images.githubusercontent.com/15225902/83355269-8fb36580-a388-11ea-9be5-10970ac83cdd.png)

I was using the latest nightly rustc.

**Comment by GuillaumeGomez** [comment]
@lzutao It's the expected behaviour: it only matches against the exact alias name.

**Comment by ollie27** [comment]
Sorry for the delay.

@rfcbot resolve use-cases

> Sometimes, to improve the API, you have to rename things and call a few functions from the C library inside a same function. Or just the renaming might be confusing sometimes (it happened a lot for me when using `rust-sdl2`).

seems reasonable

@rfcbot resolve naming

I can't think of a better name.

@rfcbot resolve associated-items

Fixed by #71724

@rfcbot resolve invalid-forms

This is mostly done, only #75888 to go.

---

There is still a weird `.replace("\"", "")` that needs to be removed or a reason for it needs to be found: https://github.com/rust-lang/rust/blob/3edf11cf811707cde49b30a80896c24189649d95/src/librustdoc/clean/types.rs#L665

I've noticed that `#![doc(alias)]` seems to be ignored when added to crate root. It might be worth making that work: #76298

> It's the expected behaviour: it only matches against the exact alias name.

I'm not sure I agree with this. I personally would expect to be able to search for aliases the same way as everything else in the search index.

I'm not aware of any other reasons to block stabilizing this though.
